File: src/main/java/com/itszuvalex/femtocraft/power/plasma/BlockPlasma.java
/*
 * ******************************************************************************
 *  * Copyright (C) 2013  Christopher Harris (Itszuvalex)
 *  * Itszuvalex@gmail.com
 *  *
 *  * This program is free software; you can redistribute it and/or
 *  * modify it under the terms of the GNU General Public License
 *  * as published by the Free Software Foundation; either version 2
 *  * of the License, or (at your option) any later version.
 *  *
 *  * This program is distributed in the hope that it will be useful,
 *  * but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  * GNU General Public License for more details.
 *  *
 *  * You should have received a copy of the GNU General Public License
 *  * along with this program; if not, write to the Free Software
 *  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *  *****************************************************************************
 */

package com.itszuvalex.femtocraft.power.plasma;

import com.itszuvalex.femtocraft.Femtocraft;
import net.minecraft.block.BlockContainer;
import net.minecraft.block.material.Material;
import net.minecraft.entity.Entity;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.AxisAlignedBB;
import net.minecraft.util.DamageSource;
import net.minecraft.util.IIcon;
import net.minecraft.world.IBlockAccess;
import net.minecraft.world.World;
import net.minecraftforge.common.util.ForgeDirection;

import java.util.List;
import java.util.Random;

/**
 * Created by Christopher Harris (Itszuvalex) on 5/6/14.
 */
public class BlockPlasma extends BlockContainer {
    public static float damagePerTick = .2f;

    @Override
    public AxisAlignedBB getCollisionBoundingBoxFromPool(World par1World, int par2, int par3, int par4) {
        return null;
    }

    public BlockPlasma() {
        super(Material.lava);
        setCreativeTab(Femtocraft.femtocraftTab());
        setBlockUnbreakable();
        setBlockTextureName(Femtocraft.ID().toLowerCase() + ":" + "BlockPlasma");
    }

    @Override
    public boolean renderAsNormalBlock() {
        return false;
    }

    @Override
    public boolean getBlocksMovement(IBlockAccess par1IBlockAccess, int par2, int par3, int par4) {
        return false;
    }


    @Override
    public IIcon getIcon(int par1, int par2) {
        return super.getIcon(par1, par2);
    }

    @Override
    public boolean isOpaqueCube() {
        return false;
    }

    @Override
    public void updateTick(World par1World, int par2, int par3, int par4, Random par5Random) {
        super.updateTick(par1World, par2, par3, par4, par5Random);
        //Tick rate is 10
        List<Entity> entities = par1World.getEntitiesWithinAABB(Entity.class,
                AxisAlignedBB.getBoundingBox
                        (par2, par3, par4, 1, 1, 1)
        );
        for (Entity entity : entities) {
            entity.attackEntityFrom(DamageSource.inFire, damagePerTick);
        }
    }

    @Override
    public int getRenderBlockPass() {
        return 1;
    }

    @Override
    public void onEntityWalking(World par1World, int par2, int par3, int par4, Entity par5Entity) {
        super.onEntityWalking(par1World, par2, par3, par4, par5Entity);
        applyEffects(par5Entity);
    }

    @Override
    public void onEntityCollidedWithBlock(World par1World, int par2, int par3, int par4, Entity par5Entity) {
        super.onEntityCollidedWithBlock(par1World, par2, par3, par4, par5Entity);
        applyEffects(par5Entity);
    }

    @Override
    public int getMobilityFlag() {
        return 2;
    }

    @Override
    public void onFallenUpon(World par1World, int par2, int par3, int par4, Entity par5Entity, float par6) {
        super.onFallenUpon(par1World, par2, par3, par4, par5Entity, par6);
        applyEffects(par5Entity);
    }

    @Override
    public boolean isBurning(IBlockAccess world, int x, int y, int z) {
        return true;
    }

    @Override
    public boolean isFireSource(World world, int x, int y, int z, ForgeDirection side) {
        return true;
    }

    @Override
    public boolean isReplaceable(IBlockAccess world, int x, int y, int z) {
        return false;
    }

    private void applyEffects(Entity entity) {
        if (!entity.isBurning()) {
            entity.setFire(4);
        }
    }

    @Override
    public void onBlockAdded(World par1World, int par2, int par3, int par4) {
        super.onBlockAdded(par1World, par2, par3, par4);
        List<Entity> entities = par1World.getEntitiesWithinAABB(Entity.class,
                AxisAlignedBB.getBoundingBox
                        (par2, par3, par4, 1, 1, 1)
        );

        for (Entity entity : entities) {
            applyEffects(entity);
        }
    }

    @Override
    public TileEntity createNewTileEntity(World world, int metadata) {
        return new TileEntityPlasma();
    }
}

File: src/main/java/com/itszuvalex/femtocraft/power/plasma/BlockPlasma.scala
/*
 * ******************************************************************************
 *  * Copyright (C) 2013  Christopher Harris (Itszuvalex)
 *  * Itszuvalex@gmail.com
 *  *
 *  * This program is free software; you can redistribute it and/or
 *  * modify it under the terms of the GNU General Public License
 *  * as published by the Free Software Foundation; either version 2
 *  * of the License, or (at your option) any later version.
 *  *
 *  * This program is distributed in the hope that it will be useful,
 *  * but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  * GNU General Public License for more details.
 *  *
 *  * You should have received a copy of the GNU General Public License
 *  * along with this program; if not, write to the Free Software
 *  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *  *****************************************************************************
 */
package com.itszuvalex.femtocraft.power.plasma

import java.util
import java.util.Random

import com.itszuvalex.femtocraft.Femtocraft
import com.itszuvalex.femtocraft.api.core.Configurable
import net.minecraft.block.BlockContainer
import net.minecraft.block.material.Material
import net.minecraft.entity.Entity
import net.minecraft.util.{AxisAlignedBB, DamageSource}
import net.minecraft.world.{IBlockAccess, World}
import net.minecraftforge.common.util.ForgeDirection

import scala.collection.JavaConversions._

/**
 * Created by Christopher Harris (Itszuvalex) on 5/6/14.
 */
object BlockPlasma {
  @Configurable val damagePerTick: Float = .2f
}

@Configurable class BlockPlasma extends BlockContainer(Material.lava) {
  setCreativeTab(Femtocraft.femtocraftTab)
  setBlockUnbreakable()
  setBlockTextureName(Femtocraft.ID.toLowerCase + ":" + "BlockPlasma")

  override def renderAsNormalBlock = false

  override def getBlocksMovement(par1IBlockAccess: IBlockAccess, par2: Int, par3: Int, par4: Int) = false

  override def getCollisionBoundingBoxFromPool(par1World: World, par2: Int, par3: Int, par4: Int) = null

  override def isOpaqueCube = false

  override def updateTick(par1World: World, par2: Int, par3: Int, par4: Int, par5Random: Random) {
    super.updateTick(par1World, par2, par3, par4, par5Random)
    par1World.getEntitiesWithinAABB(classOf[Entity], AxisAlignedBB.getBoundingBox(par2, par3, par4, 1, 1, 1)).
    asInstanceOf[util.List[Entity]].foreach(_.attackEntityFrom(DamageSource.inFire, BlockPlasma.damagePerTick))
  }

  override def getRenderBlockPass = 1

  override def onEntityWalking(par1World: World, par2: Int, par3: Int, par4: Int, par5Entity: Entity) {
    super.onEntityWalking(par1World, par2, par3, par4, par5Entity)
    applyEffects(par5Entity)
  }

  override def onEntityCollidedWithBlock(par1World: World, par2: Int, par3: Int, par4: Int, par5Entity: Entity) {
    super.onEntityCollidedWithBlock(par1World, par2, par3, par4, par5Entity)
    applyEffects(par5Entity)
  }

  private def applyEffects(entity: Entity) {
    if (!entity.isBurning) {
      entity.setFire(4)
    }
  }

  override def getMobilityFlag = 2

  override def onFallenUpon(par1World: World, par2: Int, par3: Int, par4: Int, par5Entity: Entity, par6: Float) {
    super.onFallenUpon(par1World, par2, par3, par4, par5Entity, par6)
    applyEffects(par5Entity)
  }

  override def isReplaceable(world: IBlockAccess, x: Int, y: Int, z: Int) = false

  override def isBurning(world: IBlockAccess, x: Int, y: Int, z: Int) = true

  override def isFireSource(world: World, x: Int, y: Int, z: Int, side: ForgeDirection) = true

  override def onBlockAdded(par1World: World, par2: Int, par3: Int, par4: Int) {
    super.onBlockAdded(par1World, par2, par3, par4)
    par1World.getEntitiesWithinAABB(classOf[Entity], AxisAlignedBB.getBoundingBox(par2, par3, par4, 1, 1, 1)).asInstanceOf[util.List[Entity]].
    foreach(applyEffects)
  }

  def createNewTileEntity(world: World, metadata: Int) = new TileEntityPlasma
}
